pkg/networkserver/grpc_deviceregistry.go

```diff
@@ -234,43 +234,43 @@ func (ns *NetworkServer) Get(ctx context.Context, req *ttnpb.GetEndDeviceRequest
 	return ttnpb.FilterGetEndDevice(dev, req.FieldMask.Paths...)
 }
 
-func validABPSessionKey(key *ttnpb.KeyEnvelope) bool {
-	return key != nil && key.KEKLabel == "" && !key.Key.IsZero()
-}
-
 // Set implements NsEndDeviceRegistryServer.
-func (ns *NetworkServer) Set(ctx context.Context, req *ttnpb.SetEndDeviceRequest) (*ttnpb.EndDevice, error) {
+func (ns *NetworkServer) Set(ctx context.Context, req *ttnpb.SetEndDeviceRequest) (dev *ttnpb.EndDevice, err error) {
+	if ttnpb.HasAnyField(req.FieldMask.Paths, "frequency_plan_id") && req.EndDevice.FrequencyPlanID == "" {
+		return nil, errInvalidFieldValue.WithAttributes("field", "frequency_plan_id")
+	}
+	if ttnpb.HasAnyField(req.FieldMask.Paths, "lorawan_phy_version") {
+		if err := req.EndDevice.LoRaWANPHYVersion.Validate(); err != nil {
+			return nil, errInvalidFieldValue.WithAttributes("field", "lorawan_phy_version").WithCause(err)
+		}
+	}
+	if ttnpb.HasAnyField(req.FieldMask.Paths, "lorawan_version") {
+		if err := req.EndDevice.LoRaWANVersion.Validate(); err != nil {
+			return nil, errInvalidFieldValue.WithAttributes("field", "lorawan_version").WithCause(err)
+		}
+	}
+
 	if ttnpb.HasAnyField(req.FieldMask.Paths, "session.dev_addr") && (req.EndDevice.Session == nil || req.EndDevice.Session.DevAddr.IsZero()) {
```

Comment:
`HasAnyField` also checks for prefixes, so you could eliminate multiple `req.EndDevice.Session == nil` checks below with a single

```go
if ttnpb.HasAnyField(req.FieldMask.Paths, "session") && req.EndDevice.Session == nil {
  ...
}
```

Reply:
That is not true
```
diff --git a/pkg/ttnpb/fieldmask_utils_test.go b/pkg/ttnpb/fieldmask_utils_test.go
index e51836044..fe6f3f0b8 100644
--- a/pkg/ttnpb/fieldmask_utils_test.go
+++ b/pkg/ttnpb/fieldmask_utils_test.go
@@ -101,6 +101,13 @@ func TestHasAnyField(t *testing.T) {
 	a.So(HasAnyField(requested, "x", "b.c.d"), should.BeTrue)
 	a.So(HasAnyField(requested, "d"), should.BeTrue)
 	a.So(HasAnyField(requested, "d.e", "b"), should.BeTrue)
+	a.So(HasAnyField(requested, "b"), should.BeTrue)
+	a.So(HasAnyField([]string{
+		"session.keys",
+		"session.dev_addr",
+	},
+		"session",
+	), should.BeTrue)
 }
 
 func TestFlattenPaths(t *testing.T) {
```